File: gojse/functors/sql.go
package functors

import (
	"encoding/json"
	"fmt"
	"strings"
)

// SQLFunctors contains SQL-related operators.
// NOTE: ONLY exports $query - the local operators ($pattern, $and, $*)
// are NOT globally available. They only exist within $query's local scope.
var SQLFunctors = map[string]Functor{
	"$query": query,
}

// QueryFields is the SELECT field list used by $query.
const QueryFields = "subject, predicate, object, meta"

// envHelper encapsulates environment operations needed by SQL functors.
type envHelper interface {
	EvalJSON(v interface{}) (interface{}, error)
	Load(functors map[string]Functor)
}

// parserHelper is the interface for parsing JSON into AST.
type parserHelper interface {
	Parse(v interface{}) (node interface{}, err error)
}

// nodeHelper is the interface for applying an AST node.
type nodeHelper interface {
	Apply(env interface{}) (interface{}, error)
}

// pattern generates SQL WHERE condition for a triple pattern.
// This is the LOCAL version used inside $query.
func pattern(env interface{}, args []interface{}) (interface{}, error) {
	if len(args) < 3 {
		return "", fmt.Errorf("$pattern requires (subject, predicate, object)")
	}
	subj, ok1 := args[0].(string)
	pred, ok2 := args[1].(string)
	obj, ok3 := args[2].(string)
	if !ok1 || !ok2 || !ok3 {
		return "", fmt.Errorf("$pattern requires string arguments")
	}
	triple := PatternToTriple(subj, pred, obj)
	cond, err := TripleToSQLCondition(triple)
	if err != nil {
		return "", err
	}
	// Return just the WHERE condition, not a full SELECT
	return cond, nil
}

// sqlAnd is SQL-specific AND: joins conditions with " and ".
// This is LOCAL-ONLY for $query, different from logical $and in utils.go.
func sqlAnd(env interface{}, args []interface{}) (interface{}, error) {
	envImpl, ok := env.(envHelper)
	if !ok {
		return "", fmt.Errorf("env does not implement EvalJSON")
	}

	var tokens []string
	for _, arg := range args {
		result, err := envImpl.EvalJSON(arg)
		if err != nil {
			return "", fmt.Errorf("failed to evaluate $and argument: %w", err)
		}
		sql, ok := result.(string)
		if !ok {
			return "", fmt.Errorf("$and arguments must evaluate to strings")
		}
		tokens = append(tokens, sql)
	}
	return strings.Join(tokens, " and "), nil
}

// wildcard is the wildcard helper for local scope.
func wildcard(env interface{}, args []interface{}) (interface{}, error) {
	return "*", nil
}

// localSQLFunctors returns local operators for $query scope only.
func localSQLFunctors() map[string]Functor {
	return map[string]Functor{
		"$pattern": pattern,
		"$and":     sqlAnd,
		"$*":       wildcard,
	}
}

// query generates SQL for multi-pattern query with LOCAL environment.
// Form: {"$query": condition}
// where condition is an AST expression with local operators ($pattern, $and, $*)
func query(env interface{}, args []interface{}) (interface{}, error) {
	if len(args) < 1 {
		return "", fmt.Errorf("$query expects a condition expression")
	}

	// Type assert to get the helper interfaces
	envHelper, ok := env.(envHelper)
	if !ok {
		return "", fmt.Errorf("env does not implement required interface")
	}

	// Create local evaluation context that has access to local functors
	localEnv := &localEvalContext{
		parent: envHelper,
		local:  localSQLFunctors(),
	}

	// Evaluate the condition in the local environment
	result, err := localEnv.EvalJSON(args[0])
	if err != nil {
		return "", fmt.Errorf("failed to evaluate query condition: %w", err)
	}

	whereStr, ok := result.(string)
	if !ok {
		return "", fmt.Errorf("query condition must evaluate to string, got %T", result)
	}

	sql := fmt.Sprintf(
		"select %s \nfrom statement \nwhere \n    %s \noffset 0\nlimit 100 \n",
		QueryFields,
		whereStr,
	)
	return sql, nil
}

// localEvalContext provides evaluation with local functors available.
// This simulates local scope by checking local functors first.
type localEvalContext struct {
	parent envHelper
	local  map[string]Functor
}

// EvalJSON evaluates JSON with local functors taking precedence.
func (c *localEvalContext) EvalJSON(v interface{}) (interface{}, error) {
	// Handle object expressions like {"$op": ...} or {"$quote": ...}
	if m, ok := v.(map[string]interface{}); ok && len(m) == 1 {
		for key, arg := range m {
			if key == "$quote" {
				// $quote returns its argument unevaluated
				// But we need to continue evaluating if the result is another expression
				return c.continueEval(arg)
			}
			if fn, hasLocal := c.local[key]; hasLocal {
				// Use local functor
				return c.applyFunctor(fn, arg)
			}
		}
	}

	// Handle arrays like ["$op", ...]
	if arr, ok := v.([]interface{}); ok && len(arr) > 0 {
		if key, ok := arr[0].(string); ok {
			if key == "$quote" && len(arr) > 1 {
				// $quote returns its argument unevaluated
				// But we need to continue evaluating if the result is another expression
				return c.continueEval(arr[1])
			}
			if fn, hasLocal := c.local[key]; hasLocal {
				// Use local functor with rest of array as args
				args := arr[1:]
				if len(args) == 1 {
					// Single argument - might be another expression
					return c.applyFunctor(fn, args[0])
				}
				return fn(c, args)
			}
		}
	}

	// Delegate to parent environment for everything else
	return c.parent.EvalJSON(v)
}

// continueEval continues evaluating the value until we get a non-expression result.
// This is needed after $quote to handle nested expressions.
func (c *localEvalContext) continueEval(v interface{}) (interface{}, error) {
	// If v is an expression, evaluate it and continue
	if m, ok := v.(map[string]interface{}); ok && len(m) == 1 {
		for key, arg := range m {
			if fn, hasLocal := c.local[key]; hasLocal {
				result, err := c.applyFunctor(fn, arg)
				if err != nil {
					return nil, err
				}
				// Continue evaluating if result is another expression
				return c.continueEval(result)
			}
		}
	}
	if arr, ok := v.([]interface{}); ok && len(arr) > 0 {
		if key, ok := arr[0].(string); ok {
			if fn, hasLocal := c.local[key]; hasLocal {
				args := arr[1:]
				var result interface{}
				var err error
				if len(args) == 1 {
					result, err = c.applyFunctor(fn, args[0])
				} else {
					result, err = fn(c, args)
				}
				if err != nil {
					return nil, err
				}
				// Continue evaluating if result is another expression
				return c.continueEval(result)
			}
		}
	}
	// Not an expression, return as-is
	return v, nil
}

// applyFunctor applies a functor with the given argument.
func (c *localEvalContext) applyFunctor(fn Functor, arg interface{}) (interface{}, error) {
	// Convert arg to args array form
	var args []interface{}
	if arr, ok := arg.([]interface{}); ok {
		args = arr
	} else {
		args = []interface{}{arg}
	}
	return fn(c, args)
}

// Load loads functors (for interface compatibility).
func (c *localEvalContext) Load(functors map[string]Functor) {
	// Merge into local functors, with local taking precedence
	for name, fn := range functors {
		if _, exists := c.local[name]; !exists {
			c.local[name] = fn
		}
	}
}

// PatternToTriple converts $pattern arguments into a triple slice.
func PatternToTriple(subject, predicate, object string) []string {
	if subject == "$*" && object == "$*" {
		return []string{predicate}
	}
	return []string{subject, predicate, object}
}

// TripleToSQLCondition builds a jsonb containment predicate.
func TripleToSQLCondition(triple []string) (string, error) {
	doc := map[string][]string{"triple": triple}
	data, err := json.Marshal(doc)
	if err != nil {
		return "", err
	}
	s := string(data)
	escaped := strings.ReplaceAll(s, "'", "''")
	return fmt.Sprintf("meta @> '%s'", escaped), nil
}

File: rust/src/functors/sql.rs
//! SQL extension functors for JSE.
//!
//! Following Python's functors/sql.py pattern for demonstrating local scope.
//! The SQL module ONLY exports $query, which creates a local environment
//! with local operators ($pattern, $and, $*).

use std::rc::Rc;
use std::cell::RefCell;
use serde_json::Value;
use crate::env::{Env, Functor};
use crate::ast::AstError;
use crate::ast::Parser;

pub const QUERY_FIELDS: &str = "subject, predicate, object, meta";

/// Convert $pattern arguments to PostgreSQL jsonb containment triple.
pub fn pattern_to_triple(subject: &str, predicate: &str, object: &str) -> Vec<String> {
    if subject == "*" && object == "*" {
        return vec![predicate.to_string()];
    }
    vec![
        subject.to_string(),
        predicate.to_string(),
        object.to_string(),
    ]
}

/// Build SQL WHERE clause for a triple pattern.
pub fn triple_to_sql_condition(triple: &[String]) -> String {
    let json = serde_json::json!({ "triple": triple });
    let s = json.to_string();
    let escaped = s.replace('\'', "''");
    format!("meta @> '{escaped}'")
}

/// Generate SQL WHERE condition for a triple pattern.
/// This is the LOCAL version used inside $query, which expands $* to "*".
pub fn pattern(env: &Rc<RefCell<Env>>, args: &[Value]) -> Result<Value, AstError> {
    if args.len() < 3 {
        return Err(AstError::ArityError("$pattern requires (subject, predicate, object)".to_string()));
    }

    let subj = args[0].as_str()
        .ok_or_else(|| AstError::TypeError("$pattern requires string arguments".to_string()))?;
    let pred = args[1].as_str()
        .ok_or_else(|| AstError::TypeError("$pattern requires string arguments".to_string()))?;
    let obj = args[2].as_str()
        .ok_or_else(|| AstError::TypeError("$pattern requires string arguments".to_string()))?;

    // For query's local environment, expand $* to "*" for the local scope
    let triple = pattern_to_triple(subj, pred, obj);
    let cond = triple_to_sql_condition(&triple);

    // Return just the WHERE condition, not a full SELECT
    Ok(Value::String(cond))
}

/// SQL-specific AND: joins conditions with " and ".
/// This is LOCAL-ONLY for $query, different from logical _and in utils.rs.
fn and(env: &Rc<RefCell<Env>>, args: &[Value]) -> Result<Value, AstError> {
    let mut tokens = Vec::new();
    for arg in args {
        // Parse and evaluate each argument as an expression
        let parser = Parser::new(Rc::clone(env));
        let ast = parser.parse(arg)?;
        let result = ast.apply(env)?;
        let sql = result.as_str()
            .ok_or_else(|| AstError::TypeError("$and arguments must evaluate to strings".to_string()))?;
        tokens.push(sql.to_string());
    }
    Ok(Value::String(tokens.join(" and ")))
}

/// Wildcard helper for local scope.
fn wildcard(_env: &Rc<RefCell<Env>>, _args: &[Value]) -> Result<Value, AstError> {
    Ok(Value::String("*".to_string()))
}

/// Local operators for $query scope only.
fn local_sql_functors() -> std::collections::HashMap<&'static str, Functor> {
    let mut m = std::collections::HashMap::new();
    m.insert("$pattern", pattern as Functor);
    m.insert("$and", and as Functor);
    m.insert("$*", wildcard as Functor);
    m
}

/// Generate SQL for multi-pattern query with LOCAL environment.
/// Form: [$query, condition]
/// where condition is an AST expression with local operators ($pattern, $and, $*)
pub fn query(env: &Rc<RefCell<Env>>, args: &[Value]) -> Result<Value, AstError> {
    if args.len() < 1 {
        return Err(AstError::ArityError("$query expects a condition expression".to_string()));
    }

    // Create LOCAL environment with parent
    let local = Rc::new(RefCell::new(Env::new_with_parent(Some(Rc::clone(env)))));

    // Load LOCAL operators into local scope
    local.borrow_mut().load(&local_sql_functors());

    // Create parser with LOCAL environment
    let parser = Parser::new(Rc::clone(&local));

    // Parse and evaluate in LOCAL environment
    let ast = parser.parse(&args[0])?;
    let where_clause = ast.apply(&local)?;

    // Extract the string value from the result
    let where_str = where_clause.as_str()
        .ok_or_else(|| AstError::TypeError("Query condition must evaluate to string".to_string()))?;

    // Generate SQL
    let sql = format!(
        "select {} \nfrom statement \nwhere \n    {} \noffset 0\nlimit 100 \n",
        QUERY_FIELDS,
        where_str
    );

    Ok(Value::String(sql))
}

/// SQL module functors - ONLY exports $query.
/// The local operators ($pattern, $and, $*) are NOT globally available.
/// They only exist within the local scope created by $query.
pub fn sql_functors() -> std::collections::HashMap<&'static str, Functor> {
    let mut m = std::collections::HashMap::new();
    m.insert("$query", query as Functor);
    m
}
